staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go

```diff
@@ -349,8 +411,152 @@ func (o *DebugOptions) generateDebugContainer(pod *corev1.Pod) *corev1.Ephemeral
+					Image:                    o.Image,
+					ImagePullPolicy:          o.PullPolicy,
+					Stdin:                    o.Interactive,
+					TerminationMessagePolicy: corev1.TerminationMessageReadFile,
+					TTY:                      o.TTY,
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							MountPath: "/node",
+							Name:      "node-root",
+						},
+					},
+				},
+			},
+			HostIPC:       true,
+			HostNetwork:   true,
+			HostPID:       true,
+			NodeName:      node,
+			RestartPolicy: corev1.RestartPolicyNever,
+			Volumes: []corev1.Volume{
+				{
+					Name: "node-root",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{Path: "/"},
+					},
+				},
+			},
+		},
+	}
+
+	if o.ArgsOnly {
+		p.Spec.Containers[0].Args = o.Args
+	} else {
+		p.Spec.Containers[0].Command = o.Args
+	}
+
+	return p
+}
+
+// generatePodCopy takes a Pod and returns a copy and the debug container name of that copy
+func (o *DebugOptions) generatePodCopyWithDebugContainer(pod *corev1.Pod) (*corev1.Pod, string) {
+	copied := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        o.CopyTo,
+			Namespace:   pod.Namespace,
+			Annotations: pod.Annotations,
+		},
+		Spec: *pod.Spec.DeepCopy(),
+	}
+	// change ShareProcessNamespace configuration only when commanded explicitly
+	if o.shareProcessedChanged {
+		copied.Spec.ShareProcessNamespace = pointer.BoolPtr(o.ShareProcesses)
+	}
+	if !o.SameNode {
+		copied.Spec.NodeName = ""
+	}
+
+	containerByName := containerNameToRef(copied)
+	c, containerExists := containerByName[o.Container]
+	// Add a new container if the specified container does not exist
+	if !containerExists {
+		name := o.computeDebugContainerName(copied)
+		c = &corev1.Container{Name: name}
+		// envs are customizable when adding new container
+		c.Env = o.Env
+	}
+	c.Image = o.Image
+	c.ImagePullPolicy = o.PullPolicy
+	c.Stdin = o.Interactive
+	c.TerminationMessagePolicy = corev1.TerminationMessageReadFile
+	c.TTY = o.TTY
+	if o.ArgsOnly {
+		c.Args = o.Args
+	} else {
+		c.Command = o.Args
+		c.Args = nil
+	}
+	if !containerExists {
+		copied.Spec.Containers = append(copied.Spec.Containers, *c)
+	}
+	return copied, c.Name
+}
+
+func (o *DebugOptions) computeDebugContainerName(pod *corev1.Pod) string {
+	if len(o.Container) > 0 {
+		return o.Container
+	}
+	name := o.Container
+	if len(name) == 0 {
+		cn, containerByName := "", containerNameToRef(pod)
+		for len(cn) == 0 || (containerByName[cn] != nil) {
+			cn = fmt.Sprintf("debugger-%s", nameSuffixFunc(5))
+		}
+		if !o.Quiet {
+			fmt.Fprintf(o.ErrOut, "Defaulting debug container name to %s.\n", cn)
+		}
+		name = cn
+	}
+	return name
+}
+
+func containerNameToRef(pod *corev1.Pod) map[string]*corev1.Container {
+	names := map[string]*corev1.Container{}
+	for i := range pod.Spec.Containers {
+		ref := &pod.Spec.Containers[i]
+		names[ref.Name] = ref
+	}
+	for i := range pod.Spec.InitContainers {
+		ref := &pod.Spec.Containers[i]
```

Comment:
Is there a typo here ?
In worst case, there may be index out of bounds exception.

Reply:
Hi @tedyu, this PR is based on #90094, and I think your comment is for the underlying commit. Could you follow up in #90094?